chain/network/src/concurrency/scope/mod.rs

```diff
@@ -57,254 +57,245 @@ use crate::concurrency::signal;
+        let mut m = self.0.err.lock().unwrap();
+        if m.is_some() {
+            return;
+        }
+        *m = Some(err);
+        self.0.ctx.cancel();
     }
-}
 
-impl<E: 'static + Send> Inner<E> {
-    /// Spawns a task in the scope, which owns a reference of to the scope, so that scope doesn't
-    /// terminate before all tasks are completed.
+    /// Spawns a task in the scope, which owns a reference of to the scope,
+    /// so that scope doesn't terminate before all tasks are completed.
+    ///
+    /// The reference to the scope can be either `Arc<TerminateGuard>` or `Arc<CancelGuard>`,
+    /// so that the scope may get terminated/canceled when the guard is dropped.
     ///
-    /// The reference to the scope can be an arbitrary
-    /// type, so that a custom drop() behavior can be added. For example, see `StrongService` scope reference,
-    /// which cancels the scope when dropped.
+    /// Returns a handle to the task, which awaits either for the task to return succesfully,
+    /// or for the WHOLE scope to terminate.
     fn spawn<M: 'static + Send + Sync + Borrow<Self>, T: 'static + Send>(
         m: Arc<M>,
         f: impl 'static + Send + Future<Output = Result<T, E>>,
-    ) -> tokio::task::JoinHandle<Result<T, ErrTaskCanceled>> {
+    ) -> tokio::task::JoinHandle<Result<T, Arc<Inner<E>>>> {
         tokio::spawn(must_complete(async move {
-            match (ctx::CtxFuture { ctx: m.as_ref().borrow().ctx.clone(), inner: f }).await {
+            match (ctx::CtxFuture { ctx: m.as_ref().borrow().0.ctx.clone(), inner: f }).await {
                 Ok(v) => Ok(v),
                 Err(err) => {
-                    m.as_ref().borrow().output.send(err);
-                    Err(ErrTaskCanceled)
+                    let guard = m.as_ref().borrow();
+                    guard.register(err);
+                    let inner = guard.0.clone();
+                    drop(m);
+                    inner.terminated.recv().await;
+                    Err(inner)
                 }
             }
         }))
     }
 
     /// Spawns a new service in the scope.
-    ///
-    /// A service is a scope, which gets canceled when
-    /// its handler (`Service`) is dropped. Service belongs to a scope, in a sense that
-    /// a dedicated task is spawned on the scope which awaits for service to terminate and
-    /// returns the service's result.
-    pub fn new_service(self: Arc<Self>) -> Service<E> {
-        let subscope = self.new_subscope();
-        let service = Service(Arc::downgrade(&subscope));
-        // Spawn a guard task in the Service which will prevent termination of the Service
-        // until the context is not canceled. See `Service` for a list
-        // of events canceling the Service.
-        Inner::spawn(subscope, async move { Ok(ctx::canceled().await) });
+    pub fn new_service<E2: 'static + Send>(self: Arc<Self>) -> Service<E2> {
+        let sub = Arc::new(TerminateGuard::new(&self.0.ctx));
```

Comment:
> I do recognize that the approach I'm suggesting would require holding an Arc<Scope> (instead of Weak<TerminateGuard>). This is what I'm not sure about, I don't actually understand why TerminateGuard had to be held as a Weak.

That's a very good question and actually the core of the complexity.
Scope<'env> (the public api) has to bind the lifetime 'env and it has to outlive any TerminateGuard. Therefore Scope cannot keep any Arc of TerminateGuard. Every task spawned within the Scope keeps an implicit inaccesible copy of TerminateGuard therefore the Weak<TerminateGuard> in scope can be safely upgraded within the tasks.

I understand your intution to squash these layers of abstractions and I tried that over multiple iterations. However only the separation of concerns that I introduced here: Inner (the state of the scope which can outlive the scope), TerminateGuard and CancelGuard (which statically guarantee the state of the scope), Scope (the public API of the scope module) allowed me to have actually confidence in the correctness of the implementation. The design pattern of keeping a bunch of references to the same object to guarantee a subset of properties, also turned out for me to work very well - once you adapt to this way of thinking, making changes to this code is really a breeze (which I cannot say about the previous iterations).

The "scope::Service" API however is far from final, I'm already working on the next iteration.